feat: Allow specifying a version when loading a v2 XBlock

bradenmacdonald opened the pull request:
This updates the v2 XBlock python APIs so that they can load the latest draft version of an XBlock, or the latest published, or a specific version. Before this, they could only load the draft.

## Supporting information

Working toward https://github.com/openedx/frontend-app-authoring/issues/1341

Private ref: [FAL-3879](https://tasks.opencraft.com/browse/FAL-3879)

## Testing instructions

1. Create a block that says "this is the published version". Publish it, then edit it to say "This is the draft version".
2. Open that block's embed view in both the CMS and LMS. e.g.
    - http://studio.local.openedx.io:8001/xblocks/v2/lb:OpenCraftX:ALPHA:html:12345/embed/student_view/ - studio so will show draft
    - http://local.openedx.io:8000/xblocks/v2/lb:OpenCraftX:ALPHA:html:12345/embed/student_view/ - LMS so will show published
3. With the Studio endpoint you can also pass more specific query params:
    - http://studio.local.openedx.io:8001/xblocks/v2/lb:OpenCraftX:ALPHA:html:12345/embed/student_view/?version=published - studio can be requested to show the published version
    - http://studio.local.openedx.io:8001/xblocks/v2/lb:OpenCraftX:ALPHA:html:12345/embed/student_view/?version=2 - studio can show a specific past version
    - Those are all working examples for me, but you'll need to substitute the ID of a block on your system.

## Deadline

ASAP

## TODO (in future PRs)

This PR ensures that handlers are always executed using the same version of the block that called the handler, but it doesn't yet have the same support for loading static assets from the same version of the block (because we don't have static asset support merged yet).

It's kind of awkward having to publish the entire library at once, rather than publishing just one block that you're working on. But it seems like we don't have an API to publish a single component at a time yet.

bradenmacdonald committed:
feat: Allow specifying a version when loading a v2 XBlock

openedx-webhooks commented:
Thanks for the pull request, @bradenmacdonald!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.



#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?


This repository is currently maintained by `@openedx/wg-maintenance-edx-platform`.  Tag them in a comment and let them know that your changes are ready for review.


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

bradenmacdonald committed:
fix: 500 error if a v2 library xblock checks a course waffle flag

The drag-and-drop-v2 xblock is checking the get_grading_ignore_decoys flag.

bradenmacdonald committed:
fix: v2 xblock "fields" API can now change display_name of any block

Previously it wouldn't work on blocks like drag-and-drop-v2 which didn't have XModuleMixin.

bradenmacdonald committed:
feat: Allow specifying a version when loading a v2 XBlock

bradenmacdonald committed:
test: add tests for retrieving a particular XBlock version

bradenmacdonald committed:
feat: call XBlock handlers using the same version of the block

bradenmacdonald committed:
chore: address mypy/pylint issues

bradenmacdonald committed:
test: check that handlers can't overwrite draft data with old data

DanielVZ96 commented on openedx/core/djangoapps/content_libraries/tests/test_embed_block.py:
```diff
@@ -0,0 +1,181 @@
+        html = self._embed_block(block_id, version=3)
+        check_fields('Field Test Block (Published, v3)', 'Published setting value 3.', 'Published content value 3.')
+
+        # And if we request a specific version, we get that:
+        html = self._embed_block(block_id, version=2)
+        check_fields('Field Test Block (Old, v2)', 'Old setting value 2.', 'Old content value 2.')
```
Just for exhausting all possible calls, I'd suggest adding this slightly redundant test to check that it gives you the latest draft version if you explicitly request it. 

```
html = self._embed_block(block_id, version=4)
check_fields('Field Test Block (Draft, v4)', 'Draft setting value 4.', 'Draft content value 4.')
```

DanielVZ96 commented on openedx/core/djangoapps/waffle_utils/__init__.py:
```diff
@@ -107,12 +107,12 @@ def is_enabled(self, course_key=None):  # pylint: disable=arguments-differ
+            else:
+                # In case this gets called with a content library key, that's fine - just ignore it and
+                # act like a normal waffle flag. We currently don't support library-specific overrides.
+                assert isinstance(course_key, LearningContextKey)
```
Should we keep the explanation for the assertion? 
```
            [...] "Provided course_key '{}' is not instance of CourseKey.".format(
                course_key
            )
````

pomegranited replied:
Should also update the `is_enabled` method docstring so we know it's ok to call this with a LearningContextKey | CourseKey.

bradenmacdonald replied:
Done.

DanielVZ96 commented on openedx/core/djangoapps/xblock/rest_api/views.py:
```diff
@@ -43,6 +43,25 @@
 invalid_not_found_fmt = "XBlock {usage_key} does not exist, or you don't have permission to view it."
 
 
+def parse_version_request(version_str: str | None) -> LatestVersion | int:
+    """
+    Given a version parameter from a query string (?version=14, ?version=draft,
+    ?version=published), get the LatestVersion parameter to use with the API.
+    """
+    if version_str is None:
+        return LatestVersion.AUTO  # AUTO = published if we're in the LMS, draft if we're in Studio.
+    if version_str == "draft":
+        return LatestVersion.DRAFT
+    if version_str == "published":
+        return LatestVersion.PUBLISHED
+    try:
+        return int(version_str)
+    except ValueError:
+        raise serializers.ValidationError(  # pylint: disable=raise-missing-from
+            "Invalid version specifier '{version_str}'. Expected 'draft', 'published', or an integer."
+        )
```
nitpick: not a fan of hungarian notation. not sure if there's value in using it here.
```suggestion
def parse_version_request(version: str | None) -> LatestVersion | int:
    """
    Given a version parameter from a query string (?version=14, ?version=draft,
    ?version=published), get the LatestVersion parameter to use with the API.
    """
    if version is None:
        return LatestVersion.AUTO  # AUTO = published if we're in the LMS, draft if we're in Studio.
    if version == "draft":
        return LatestVersion.DRAFT
    if version == "published":
        return LatestVersion.PUBLISHED
    try:
        return int(version)
    except ValueError:
        raise serializers.ValidationError(  # pylint: disable=raise-missing-from
            "Invalid version specifier '{version}'. Expected 'draft', 'published', or an integer."
        )
```

bradenmacdonald replied:
In some parts of the code we have `version` with type `LatestVersion | int` so I thought it's confusing if other parts of the code have `version` with type `str | None` which is quite different. I could call it `version_request` if that satisfies your aversion to Hungarian notation. But I don't have strong feelings at all, just want to do what's least confusing for the reader.

DanielVZ96 replied:
ohh that's a good reason, so I have no issues with keeping it

DanielVZ96 reviewed (commented):
LGTM in general. Just left some small comments

ormsbee commented on openedx/core/djangoapps/xblock/rest_api/views.py:
```diff
@@ -43,6 +43,25 @@
 
+def parse_version_request(version_str: str | None) -> LatestVersion | int:
+    """
+    Given a version parameter from a query string (?version=14, ?version=draft,
+    ?version=published), get the LatestVersion parameter to use with the API.
```
Does it make sense to make this an optional part of the URL pattern matching instead of a querystring? So something like:`xblocks/v2/<str:usage_key_str>;<str:version>/`. I mostly ask because it seems like this convention could eventually make its way into other calls that could have their own querystring handling, e.g. XBlock handler calls.

bradenmacdonald replied:
Hah, that's almost exactly what I had in an earlier version of this PR and for that exact reason. However, I later saw that the [XBlock handler APIs can account for query strings already being in the handler URL](https://github.com/openedx/edx-platform/blob/5b809673b6162a1962ab7f9ca11ec9219e422d92/openedx/core/djangoapps/xblock/runtime/runtime.py#L146-L148) so it seemed like doing it that way in some cases just complicated things, and using `?version` in the query string was simpler and more consistent.

![diff](https://github.com/user-attachments/assets/24aad119-4f9a-438b-910e-af3952270732)

bradenmacdonald replied:
@ormsbee Please let me know if that makes sense ^ and if you wanna take another look at this PR before I merge.

ormsbee replied:
@bradenmacdonald: Okay, I think that's fine. And if it becomes a problem, we can address it then. No need to re-review on my part. Thank you!

bradenmacdonald replied:
@ormsbee You were right 😅

https://github.com/openedx/edx-platform/issues/35673

pomegranited commented on openedx/core/djangoapps/xblock/runtime/learning_core_runtime.py:
```diff
@@ -224,6 +239,13 @@ def save_block(self, block):
         if not self.authored_data_store.has_changes(block):
             return  # No changes, so no action needed.
 
+        if block._runtime_requested_version != LatestVersion.DRAFT:  # pylint: disable=protected-access
```
If we've loaded the block with a numeric version instead of `LatestVersion.DRAFT`, this check will fail, even if that version number matches the latest draft's version number.

bradenmacdonald replied:
I know. But generally if you're loading a specific version, I assume you're not planning to make changes. If you were planning to make changes, you'd always want to load LatestVersion.DRAFT to avoid race conditions.

pomegranited reviewed (approved):
:+1: This is working perfectly -- so clean how the plugin system lets us use different defaults for xblock.app LMS vs CMS.

Note for a future PR to Authoring MFE: the Component sidebar uses the CMS "draft" version for editing the title, but the LMS "published" version for the Preview:

![image](https://github.com/user-attachments/assets/80127d2e-4ee4-4407-a9b9-be659d5e4724)

- [x] I tested this using the PR test instructions.
- [x] I read through the code -- noted potential issue in `save_block`
- [x] I checked for accessibility issues
- [x] Includes documentation -- noted `is_enabled` method docstring needs updating
- [x] <strike>User-facing strings are extracted for translation</strike> N/A

kdmccormick commented:
Removing my review request since you already have 3 solid reviews 🚀

bradenmacdonald committed:
docs: address review comments
